appscale/tools/agents/azure_agent.py

```diff
@@ -0,0 +1,694 @@
+
+    # Sleep until an IP address gets associated with the VM.
+    sleep_time = 1
+    while True:
+      public_ip_address = network_client.public_ip_addresses.get(resource_group,
+                                                                 vm_network_name)
+      if public_ip_address.ip_address:
+        print('Azure VM is available at {}'.format(public_ip_address.ip_address))
+        break
+      AppScaleLogger.verbose("Waiting {} second(s) for IP address to be "
+        "available".format(sleep_time), verbose)
+      time.sleep(sleep_time)
+      sleep_time = min(sleep_time * 2, 20)
+
+  def associate_static_ip(self, instance_id, static_ip):
+    """ Associates the given static IP address with the given instance ID.
+
+    Args:
+      instance_id: A str that names the instance that the static IP should be
+        bound to.
+      static_ip: A str naming the static IP to bind to the given instance.
+    """
+
+  def terminate_instances(self, parameters):
+    """ Deletes the instances specified in 'parameters' running in Azure.
+    Args:
+      parameters: A dict, containing all the parameters necessary to
+        authenticate this user with Azure.
+    """
+    credentials = self.open_connection(parameters)
+    resource_group = parameters[self.PARAM_RESOURCE_GROUP]
+    subscription_id = parameters[self.PARAM_SUBCR_ID]
+    public_ips, private_ips, instance_ids = self.describe_instances(parameters)
+
+    AppScaleLogger.log("Terminating the vm instance/s '{}'".format(instance_ids))
+    compute_client = ComputeManagementClient(credentials, subscription_id)
+    for vm_name in instance_ids:
+      result = compute_client.virtual_machines.delete(resource_group, vm_name)
+      resource_name  = 'Virtual Machine' + ':' + vm_name
+      self.sleep_until_delete_operation_done(result, resource_name)
+
+  def sleep_until_delete_operation_done(self, result, resource_name):
+    """ Sleeps until the delete operation for the resource is completed
+    successfully.
+    Args:
+      result: An instance, of the AzureOperationPoller to poll for the status
+      of the operation being performed.
+      resource_name: The name of the resource being deleted.
+    """
+    sleep_time = 1
+    while not result.done():
+      AppScaleLogger.log("Waiting {0} second(s) for '{1}' to be "
+                         "deleted".format(sleep_time, resource_name))
+      time.sleep(sleep_time)
+      sleep_time = min(sleep_time * 2, 20)
+
+  def does_address_exist(self, parameters):
+    """ Verifies that the specified static IP address has been allocated, and
+    belongs to the user with the given credentials.
+
+    Args:
+      parameters: A dict containing values necessary to authenticate with the
+        underlying cloud, as well as a key indicating which static IP address
+        should be validated.
+    Returns:
+      A bool that indicates if the given static IP address exists, and belongs
+      to this user.
+    """
+
```

Comment:
I have not currently looked into static ip allocation for Azure. So this method body is empty, but I will look into accounting for this in the next release.